cdap-master/src/main/java/co/cask/cdap/data/runtime/main/StartupChecker.java

```diff
@@ -0,0 +1,179 @@
+   * are handled in a separate method.
+   *
+   * This method is a no-op if {@link Constants#STARTUP_CHECKS_ENABLED} is false.
+   *
+   * @return whether the sanity checks succeeded or not
+   */
+  public boolean runSanityChecks() {
+    boolean noErrors = true;
+
+    ConfigurationLogger.logImportantConfig(cConf);
+    if (cConf.getBoolean(Constants.STARTUP_CHECKS_ENABLED)) {
+      noErrors = checkHDFS();
+      noErrors = checkYarn() && noErrors;
+      noErrors = checkHBase() && noErrors;
+    }
+
+    return noErrors;
+  }
+
+  private boolean checkHDFS() {
+    LOG.info("Checking that HDFS is available.");
+    FileSystem hdfs;
+    try {
+      hdfs = FileSystem.get(hConf);
+    } catch (Exception e) {
+      LOG.error("Unable to get an HDFS FileSystem instance. " +
+                  "Please check that the correct Hadoop libraries and configuration are included in the classpath.", e);
+      return false;
+    }
+
+    try {
+      hdfs.getStatus();
+    } catch (Exception e) {
+      LOG.error("Unable to get the status of HDFS. Please check that HDFS is running, " +
+                  "and that the correct Hadoop libraries and configuration are included in the classpath.", e);
+      return false;
+    }
+
+    return checkHDFSPermissions();
+  }
+
+  private boolean checkHDFSPermissions() {
+    String hdfsUser = cConf.get(Constants.CFG_HDFS_USER);
+    String hdfsRootPath = cConf.get(Constants.CFG_HDFS_NAMESPACE);
+    Location rootLocation = locationFactory.create(hdfsRootPath);
+    try {
+      if (!rootLocation.mkdirs()) {
+        LOG.error("Could not make directory {} on hdfs. " +
+                    "Please check that user {} has permission to write to {}, " +
+                    "or create the directory manually with write permissions.",
+                  hdfsRootPath, hdfsUser, hdfsRootPath);
+        return false;
+      }
+    } catch (Exception e) {
+      LOG.error("Error while ensuring that directory {} exists on hdfs with the correct permissions. " +
+                  "Please check that HDFS is running and that user {} has permission to write to {}, " +
+                  "or create the directory manually with write permissions.",
+                hdfsRootPath, hdfsUser, hdfsRootPath, e);
+      return false;
+    }
+    return true;
+  }
+
+  private boolean checkHBase() {
+    LOG.info("Checking that HBase is available.");
+    try (final HConnection hbaseConnection = HConnectionManager.createConnection(hConf)) {
+      hbaseConnection.listTables();
+      return true;
+    } catch (IOException e) {
+      LOG.error("Error connecting to HBase. " +
+                  "Please check that HBase is running and that the correct Hadoop libraries and configuration " +
+                  "are included in the CDAP master classpath.", e);
+      return false;
+    }
+  }
+
+  private boolean checkYarn() {
+    LOG.info("Checking that YARN is available.");
```

Comment:
You can say in the message that this can take upto 60 seconds depending on status of YARN. That will be more informative -- otherwise user will not know what's happening.